### DISCUSSION TITLE
Doxygen Comments for Python Functions with **kwargs

### DISCUSSION THREAD

**Discussion by pipsqik** [swig/swig#2266] [General]
Hi everybody

When using the *-keyword* option and Swig needs to create a Python function that uses **kwargs, the docstring will keep the original c++ parameter names.   Eg the C++ code:
```
/**
\param a  Some Comment About an Integer a
\param b [optional] vector....
\param c [optional] string...
*/
void foo( int a, std::vector<std::string> b = std::vector<std::string>(), std::string c = "")
```

would be converted to a python function:
```
def foo(*args, **kwargs)
r"""
    :type a: int
    :param a: Some Comment About an Integer a
    :type b: list[str]
    :param b:  [optional] vector ....
    :type c: str
    :param c: [optional] string ...
"""
```

PyCharm in particular and possibly other Python IDEs will not display this docstring as it checks the parameter names.  Only 'args' and 'kwargs' are actual parameters so the comments for 'a', 'b', and 'c' are not shown.   Instead PyCharm requires the 'keyword' specifier.  IE in the above example the Python code should look like:
```
def foo(*args, **kwargs)
r"""
    :type a: int
    :keyword a: Some Comment About an Integer a
    :type b: list[str]
    :keyword b:  [optional] vector ....
    :type c: str
    :keyword c: [optional] string ...
"""
```

(In fact, in this example the 'type' specifiers are irrelevant as PyCharm currently has no way of display the types of keyword arguments.  But that seems like it should be fixed by JetBrains...)

Is this change, something that others would find useful for their development environments?  Is anyone currently looking at it?

regards

**Comment by wsfulton** [comment]
I believe a Doxygen 'param' is output as a PyDoc 'param' by design and works with Sphinx. If you can provide a link to what is officially supported by Sphinx, I'd expect it would be easy to create a pull request with this mapping change.

**Comment by pipsqik** [comment]
Hi, 

The only documentation I can find that alludes to this is for the sphinx napoleon extension:
[https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy)
Both the numpy and google style documentation says:
`If *args or **kwargs are accepted,
    they should be listed as '*args' and '**kwargs'.`

And it seems editors like PyCharm are enforcing this.  PyCharm and perhaps some other Python editors will not display the parameter comments when the named keywords are not ACTUAL parameters of the function. In the example above the actual parameters of the Python function are '*args' and '**kwargs'.  So the swig-generated comments about parameters a, b, and c do not get rendered to the PyCharm help window.  They are there in the docstring for people who view the raw docstring, but not rendered by the PyCharm editor.

Then the only option is to have a docstring entry `:param **kwargs: ...` and list all the keyword arguments in the one comment, or as I noted above to use, for example `:keyword a: comment about parameter a...`.   I cannot find any specific documentation about the use of the `:keyword` identifier, but I can vouch that it is rendered correctly on PyCharm (and Spyder).
